Update experiment setup script for free-forecast ensemble

XianwuXue-NOAA opened the pull request:
<!-- PLEASE READ -->
<!--
Before opening a PR, please note these guidelines:

- Each PR should only address ONE topic and have an associated issue
- No hardcoded or paths to personal directories should be present
- No temporary or backup files should be committed
- Any code that was disabled by being commented out should be removed
-->

**Description**

As we move towards the merger of GEFS into the GFS system, we need to be able to generate experiments using setup_expt.py with an arbitrary number of free-forecast ensemble members in addition to the number of EnKF members.

<!-- Please include relevant motivation and context. -->
<!-- Please include a summary of the change and which issue is fixed. -->
<!-- List any dependencies that are required for this change. -->

<!-- Please provide reference to the issue this pull request is addressing. -->
<!-- For e.g. Fixes #IssueNumber -->

**Type of change**

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update (add "gefs" to setup_expt.py)

**How Has This Been Tested?**

Tested on Hera for ATM only GEFS forecast-only

<!-- Please describe the tests that you ran to verify your changes and on the platforms these tests were conducted. -->
<!-- Provide instructions so we can reproduce. -->
<!-- Please also list any relevant details for your test configuration -->

<!-- Use the following as a guide to list your tests and delete options that are not relevant. Expand as necessary. -->
<!--
- [ ] Clone and Build tests on WCOSS Dell P3
- [ ] Cycled test on Orion
- [ ] Forecast-only test on Hera
-->
  
**Checklist**

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing tests pass with my changes

XianwuXue-NOAA committed:
Update experiment setup script for free-forecast ensemble

 On branch feature/gefs_v13_789_setupexp
	modified:   parm/config/config.base.emc.dyn
	new file:   parm/config/config.gefs
	modified:   ush/rocoto/setup_expt.py
	modified:   ush/rocoto/workflow_utils.py

Refs: #789

XianwuXue-NOAA committed:
add sorc/upp.fd to .gitignore

 On branch feature/gefs_v13_789_setupexp
	modified:   .gitignore

Refs: #789

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/gefs_v13_789_setupexp

WalterKolczynski-NOAA commented on ush/rocoto/setup_expt.py:
```diff
@@ -30,7 +30,8 @@ def fill_COMROT(host, inputs):
         'cycled': fill_COMROT_cycled,
-        'forecast-only': fill_COMROT_forecasts
+        'forecast-only': fill_COMROT_forecasts,
+        'gefs': fill_COMROT_gefs
```
Instead of adding a new mode, there should be a new command line option to specify the number of ensemble members (with a default of 0). While we are starting with forecast-only, at some point in the future we will want to run full cycling and also have a forecast ensemble.

WalterKolczynski-NOAA replied:
Actually, we might want to use -1 as the default (meaning high-res only). For now assume GEFS-resolution only if the number of members >= 0, though we'll probably need a new switch in the future to turn off the high-res instead, since we'll want the option to everything (GFS, GEFS, GDAS, EnKF) together at some point, but we don't always want to run high-res when we run GEFS.

WalterKolczynski-NOAA commented on ush/rocoto/setup_expt.py:
```diff
@@ -231,18 +241,21 @@ def input_args():
-                        type=int, required=False, default=192)
-    cycled.add_argument('--nens', help='number of ensemble members',
-                        type=int, required=False, default=20)
+    for subp in [cycled, gefs]:
```
New options need to be created for the forecast ensemble rather than overloading the options that were created for EnKF. Existing options should be renamed `--nens-enkf` and `--res-enkf` to avoid confusion.

XianwuXue-NOAA replied:
OK. I will add --nens-gefs, --res-gefs to setup_expt.py and also make some changes in config.gefs.

XianwuXue-NOAA replied:
How about the changes below:

    if inputs.mode in ['cycled']:
        extend_dict = {
            "@CASEENS@": f'C{inputs.resens}',
            "@NMEM_ENKF@": inputs.nens,
        }
    elif inputs.mode in ['gefs']:
        extend_dict = {
            "@CASEENS_GEFS@": f'C{inputs.resens}',
            "@NMEM_GEFS@": inputs.nens,
        }
    elif inputs.mode in ['forecast-only', 'gefs']:
        extend_dict = {
            "@DO_AERO@": inputs.aerosols,
        }

WalterKolczynski-NOAA commented on ush/rocoto/setup_expt.py:
```diff
@@ -147,12 +155,12 @@ def edit_baseconfig(host, inputs):
-    if inputs.mode in ['cycled']:
+    if inputs.mode in ['cycled', 'gefs']:
         extend_dict = {
             "@CASEENS@": f'C{inputs.resens}',
```
This variable name will need to be changed in config.base and everywhere downstream to something like `$CASE_ENKF` to avoid confusion with new resolution variable for GEFS.

XianwuXue-NOAA replied:
How about I change to, however, I keep inputs.resens and inputs.nens, then users don't need to change command options

    if inputs.mode in ['cycled']:
        extend_dict = {
            "@CASEENS@": f'C{inputs.resens}',
            "@NMEM_ENKF@": inputs.nens,
        }
    elif inputs.mode in ['gefs']:
        extend_dict = {
            "@CASEENS_GEFS@": f'C{inputs.resens}',
            "@NMEM_GEFS@": inputs.nens,
        }
    elif inputs.mode in ['forecast-only', 'gefs']:
        extend_dict = {
            "@DO_AERO@": inputs.aerosols,
        }

WalterKolczynski-NOAA replied:
This suggested code won't even work properly, and I still want to update the options so it is clear they belong to EnKF. As always when we change a script signature, we will make an announcement when the change is merged so everyone is aware.

WalterKolczynski-NOAA replied:
Plus, I still don't want to add gefs as a new mode.

XianwuXue-NOAA replied:
If you don't want to add gefs as a new mode, do you have any suggestions?

> This suggested code won't even work properly, and I still want to update the options so it is clear they belong to EnKF. As always when we change a script signature, we will make an announcement when the change is merged so everyone is aware.

Yes, I should also change the configure file. If you agree these changes, I will work on it.

XianwuXue-NOAA replied:
> Plus, I still don't want to add gefs as a new mode.

I think adding gefs as a new mode is clear. Any suggestions?

WalterKolczynski-NOAA replied:
Yes, it's in the first change request (along with why)

XianwuXue-NOAA replied:
Originally, I want to reuse the existing variables. However, based on this conversation, I changed my mind to add new variables. If I misunderstood, please let me know

WalterKolczynski-NOAA reviewed (changes requested):
This needs to be reworked to be more inline with what we discussed previously (unless I'm forgetting or missed a conversation).

Also, the PR title and description need a *lot* of work, as they are practically non-existent right now.

XianwuXue-NOAA committed:
Add options (resens_gefs and nens_gefs) and modify resens and nens

1) Add options (resens_gefs and nens_gefs) for GEFS
2) To make it clear, modify "resens" and "nens" to
    "resens_enkf" and "nens_enkf" for ENKF

 On branch feature/gefs_v13_789_setupexp
	modified:   parm/config/config.base.emc.dyn
	modified:   ush/rocoto/setup_expt.py
Refs: #789

XianwuXue-NOAA commented:
@WalterKolczynski-NOAA I have modified the variables and also added new options for GEFS.

XianwuXue-NOAA committed:
Change the default value of option nens_gefs to ZERO

 On branch feature/gefs_v13_789_setupexp
	modified:   ush/rocoto/setup_expt.py

Refs: #789

aerorahul commented on .gitignore:
```diff
@@ -122,7 +122,7 @@ sorc/wafs_gcip.fd
 sorc/wafs_makewafs.fd
 sorc/wafs_setmissing.fd
-
+sorc/upp.fd
```
how is this related to this PR?

XianwuXue-NOAA replied:
No relate to this PR, so I will delete it.

aerorahul commented on parm/config/config.gefs:
```diff
@@ -0,0 +1,11 @@
+#!/bin/ksh -x
```
Which job is sourcing this file?
It may be needed in subsequent PR's.
Consider adding when it is used.

XianwuXue-NOAA replied:
Currently, this config file is just for placeholder. If there is no need for now until we need this file, I can delete this file.

aerorahul commented on ush/rocoto/workflow_utils.py:
```diff
@@ -272,6 +272,9 @@ def source_configs(configs, tasks):
         elif 'wave' in task:
             files.append(find_config(f'config.wave', configs))
             files.append(find_config(f'config.{task}', configs))
+        elif 'gefs' in task:
+            files.append(find_config(f'config.gefs', configs))
+            files.append(find_config(f'config.{task}', configs))
```
There are no gefs tasks currently, so this is a placeholder.
Since there is work ongoing on refactoring setup_workflow, this could be held till actual tasks are being added.

XianwuXue-NOAA replied:
Currently, I use the existing getic, init and fcst tasks for GEFS, later, I will add recenter task for GEFS. OK, I can delete these lines until actual tasks are being added

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -224,6 +225,8 @@ if [[ "$CDUMP" == "gdas" ]] ; then
    export FTSFS=10
 elif [[ "$CDUMP" == "gfs" ]] ; then
    export FHCYC=24
+elif [[ "$CDUMP" == "gefs" ]] ; then
+   export FHCYC=24
```
May be replace lines 223-230 with `case ${CDUMP} in` statements and condense `gfs` and `gefs`

XianwuXue-NOAA replied:
OK. I will do it.

aerorahul commented on parm/config/config.base.emc.dyn:
```diff
@@ -207,7 +207,8 @@ export OPS_RES="C768" # Do not change
 export LEVS=128
 export CASE="@CASECTL@"
-export CASE_ENKF="@CASEENS@"
+export CASE_ENKF="@CASEENS_ENKF@"
+export CASE_GEFS="@CASEENS_GEFS@"
```
I am still not convinced we will have a use case for running GEFS that is a resolution different than the cycled ensemble resolution.

XianwuXue-NOAA replied:
@aerorahul Any suggestions? Should I change it back to 'export CASE_ENKF="@CASEENS@"' and the corresponding changes for the options or keep this changes?

aerorahul reviewed (changes requested):
please see comments.

XianwuXue-NOAA committed:
Remove unrelated change

 On branch feature/gefs_v13_789_setupexp
	modified:   .gitignore

Refs: #789

XianwuXue-NOAA committed:
Change if to case

 On branch feature/gefs_v13_789_setupexp
	modified:   parm/config/config.base.emc.dyn

Refs: #789

XianwuXue-NOAA committed:
Change back to resens and nens because GEFS may
    use the same resolution with ENKF

 On branch feature/gefs_v13_789_setupexp
	modified:   parm/config/config.base.emc.dyn
	modified:   ush/rocoto/setup_expt.py

Refs: #789

XianwuXue-NOAA committed:
Fix some typo
 On branch feature/gefs_v13_789_setupexp
 Changes to be committed:
	modified:   parm/config/config.base.emc.dyn
	modified:   ush/rocoto/setup_expt.py

Refs: #789

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/gefs_v13_789_setupexp

 Conflicts:
	ush/rocoto/setup_expt.py # solved

 On branch feature/gefs_v13_789_setupexp

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/gefs_v13_789_setupexp

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/gefs_v13_789_setupexp

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/gefs_v13_789_setupexp

 Conflicts:
	ush/rocoto/setup_expt.py
	ush/rocoto/workflow_utils.py

 On branch feature/gefs_v13_789_setupexp
 Changes to be committed:
	modified:   .gitignore
	modified:   parm/config/config.base.emc.dyn
	modified:   parm/config/config.base.nco.static
	modified:   parm/config/config.fv3
	modified:   parm/config/config.resources
	modified:   parm/config/config.wave
	modified:   sorc/checkout.sh
	new file:   ush/ecflow/README.md
	new file:   ush/ecflow/__init.py__
	new file:   ush/ecflow/ecflow_build.yml
	new file:   ush/ecflow/ecflow_setup/__init.py__
	new file:   ush/ecflow/ecflow_setup/ecflow_definitions.py
	new file:   ush/ecflow/ecflow_setup/ecflow_setup.py
	new file:   ush/ecflow/prod.yml
	new file:   ush/ecflow/setup_workflow.py
	new file:   ush/rocoto/applications.py
	new file:   ush/rocoto/configuration.py
	deleted:    ush/rocoto/fv3gfs_workflow.sh
	new file:   ush/rocoto/hosts.py
	new file:   ush/rocoto/hosts/hera.yaml
	new file:   ush/rocoto/hosts/orion.yaml
	new file:   ush/rocoto/hosts/wcoss_dell_p3.yaml
	modified:   ush/rocoto/rocoto.py
	modified:   ush/rocoto/setup_expt.py
	deleted:    ush/rocoto/setup_workflow.py
	deleted:    ush/rocoto/setup_workflow_fcstonly.py
	new file:   ush/rocoto/setup_xml.py
	new file:   ush/rocoto/test_configuration.py
	new file:   ush/rocoto/test_hosts.py
	new file:   ush/rocoto/workflow_tasks.py
	deleted:    ush/rocoto/workflow_utils.py
	new file:   ush/rocoto/workflow_xml.py

Refs: #789

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/gefs_v13_789_setupexp

XianwuXue-NOAA committed:
Modify NET to pass value from rocoto

 On branch feature/gefs_v13_789_setupexp
	modified:   parm/config/config.base.emc.dyn

Refs: #789

XianwuXue-NOAA committed:
Remove the changes in .gitignore

 On branch feature/gefs_v13_789_setupexp
	modified:   .gitignore

Refs: #789

XianwuXue-NOAA committed:
Fix a typo

 On branch feature/gefs_v13_789_setupexp
	modified:   parm/config/config.base.emc.dyn

Refs: #789

XianwuXue-NOAA committed:
Merge branch 'develop' into feature/gefs_v13_789_setupexp
